doc/manual/contract.md

```diff
@@ -0,0 +1,505 @@
+	is
+		ensure(old(position.clone) + direction  == position)
+	do
+		position =  position + direction
+	end
+end
+~~~
+
+Note that it is often better to capture only the interesting part, rather than providing a duplication method. The duplication method is generally reserved when you deal with complex objects.
+
+## Writing invariants
+
+Invariants are used to specify the characteristics of a class/interface which must always be true, except when executing a member function. In other words, invariants define conditions that must be respected by our instance. Any instance which does not respect one or more of those conditions will be considered as in an incoherent state, and will cause an assertion failed.
+
+To define an invariant we use the class annotation `invariant`. Here is an example with our Stack. The class defines that it is impossible to be in a state where the size is less than 0.
+
+~~~
+class Stack
+	invariant(size >= 0)
+end
+~~~
+
+Another example with a class that represents a date (day, month). To be in a consistent state, the date must be represented with a day between 1 and 31 and a month between 1 and 12.
+
+~~~
+class Date
+	invariant(day >= 1 and day <= 31)
+	invariant(month >= 1 and month <= 12)
+
+	var day: Int
+	var month: Int
+end
+~~~
+
+When a class defines an invariant all methods and constructors (inherited, redefined and introduced) will check this one. By default, the invariants are only checked on exit. See section `Verification policy` to activate them on in and out. For all constructors the invariant is only checked at the end of this, this is due to the fact that the object is not yet initialized. See the section `Execution order` for example.
+
+The invariant is not checked when the calle is on `self`. The following example shows the execution of invariant:
+
+~~~
+class A
+	invariant(print_invariant)
+
+	fun print_invariant: Bool
+	do
+		print "In invariant verification"
+		return true
+	end
+
+	fun foo
+	do
+		print "In foo"
+		self.bar # The called property is a member of "self", the invariant will not be checked.
+	end
+
+	fun bar
+	do
+		print "In bar"
+	end
+end
+
+var a = new A
+a.foo # The called property is not a member of "self", the invariant will be checked.
+~~~
+
+This example gives the following output:
+~~~
+In invariant verification
+In foo
+In bar
+In invariant verification
+~~~
+
+## Inheritance
+
+The Nit contracts fully support multiple inheritance and refinement. The specialization can be interpreted as subcontracting. All methods, class/interface which are redefined must keep the set of conditions defined by all the previous definitions. Here are the subtyping rules for each type of the contracts:
+
+* The `expect` can be weakened, which guarantees that all the contracts introduced in the previous definitions will always be considered as valid entries. During the specialization of a method it will be possible to redefine its precondition in order the widen the allowed input.
```

Comment:
```suggestion
* The `expect` can be weakened, which guarantees that all the contracts introduced in the previous definitions will always be considered as valid entries. During the specialization of a method it will be possible to redefine its preconditions in order the widen the allowed input.
```